### ISSUE TITLE
[Roadmap] XGBoost 1.0.0 Roadmap

### ISSUE THREAD

**Issue by CodingCat** [dmlc/xgboost#4680]
@dmlc/xgboost-committer please add your items here by editing this post. Let's ensure that 

* each item has to be associated with a ticket 

* major design/refactoring are associated with a RFC before committing the code 

* blocking issue must be marked as blocking

* breaking change must be marked as breaking

for other contributors who have no permission to edit the post, please comment here about what you think should be in 1.0.0

I have created three new types labels, 1.0.0, Blocking, Breaking

- [x] Improve installation experience on Mac OSX (#4477)
- [x] Remove old GPU objectives.
- [x]  Remove gpu_exact updater (deprecated) #4527
- [x]  Remove multi threaded multi gpu support (deprecated) #4531
- [x]  External memory for gpu and associated dmatrix refactoring #4357 #4354
- [ ] Spark Checkpoint Performance Improvement (https://github.com/dmlc/xgboost/issues/3946) 
- [x] [BLOCKING] the sync mechanism in hist method in master branch is broken due to the inconsistent shape of tree in different workers (https://github.com/dmlc/xgboost/pull/4716, https://github.com/dmlc/xgboost/issues/4679)
- [x] Per-node sync slows down distributed training with 'hist' (#4679)
- [x] Regression tests including binary IO compatibility, output stability, performance regressions.

**Comment by thesuperzapper** [comment]
Not a committer, but can we please target PySpark API for 1.0? 
Issue: #3370 
Current PR: #4656

**Comment by CodingCat** [comment]
for other contributors who have no permission to edit the post, please comment here about what you think should be in 1.0.0

**Comment by thesuperzapper** [comment]
Also, should we target moving exclusively to the Scala based Rabit tracker (for Spark) in 1.0?

**Comment by trams** [comment]
I am also not a committer but me and the company I work in is very interested in fixing the performance issue with checkpointing (or at least mitigate it) #3946

**Comment by thesuperzapper** [comment]
There is also the secondary consideration, that we might not be ready for 1.0, and the API guarantees that come with that, for example, we could instead do 0.10.0 next?

**Comment by RAMitchell** [comment]
Added some gpu related items.

**Comment by chenqin** [comment]
would like to get native xgb fix included.
https://github.com/dmlc/xgboost/issues/4753

**Comment by trivialfis** [comment]
JSON is removed from the list.  See https://github.com/dmlc/xgboost/pull/4683#issuecomment-520485615

**Comment by thesuperzapper** [comment]
I raised an issue for my above suggestion: #4781 (To remove the python Rabit tracker)

**Comment by Daniel8hen** [comment]
FeatureImportance in the Spark version will be great as well (i.e. easily have the feature Importance)
https://github.com/dmlc/xgboost/pull/988

**Comment by trivialfis** [comment]
Added regression test.

**Comment by hcho3** [comment]
@chenqin I'd like to hear from you about regression tests, since you have experience with managing ML in production. Any suggestions?

**Comment by chenqin** [comment]
> @chenqin I'd like to hear from you about regression tests, since you have experience with managing ML in production. Any suggestions?

I think we should cover regression test on various of workloads and benchmark against prediction accuracy and stability (equal or better) than previous version within approximate same time. Two candidates on top of my head are

https://archive.ics.uci.edu/ml/datasets/HIGGS

sparse Dmatrix
https://www.kaggle.com/c/ClaimPredictionChallenge

We can try various of tree methods and configurations to ensure good coverage

tree_method, configurations / dataset / standalone or cluster

**Comment by thesuperzapper** [comment]
We should add fixing #4779 as a requirement to ship 1.0

**Comment by codingforfun** [comment]
I add  #4899 as a cleanup step.

**Comment by hcho3** [comment]
@dmlc/xgboost-committer Since we have quite a few tasks left for 1.0, maybe we should release an interim release 0.91?

**Comment by thesuperzapper** [comment]
Or perhaps 0.10.0

**Comment by trivialfis** [comment]
@thesuperzapper That will confuse version system.  I don't mind a 0.91 release, but still I want to see proper procedures for regression tests.

**Comment by thesuperzapper** [comment]
@trivialfis If master has API changes, shouldn't we bump a major version, which I guess would look like 0.100.0

**Comment by hcho3** [comment]
@thesuperzapper The 1.0.0 version is the first version we would adopt semantic versioning scheme, so no, semantic versioning won't apply to the interim release.

**Comment by CodingCat** [comment]
If we want a 0.91, we should review all changes and ensure that 0.91 is an
incremental update based on 0.90, and as such, we don’t hurt our roadmap of
1.0.0 by shifting several features to 0.9x or any other version

My suggestion would be release 1.0.0.preview.1, some other project also
does this before a major release

**Comment by hcho3** [comment]
@CodingCat 1.0.0.preview.1 is an interesting suggestion. Does Maven accept this version?

**Comment by CodingCat** [comment]
yes, you can have non-numeric letters in version number

**Comment by RAMitchell** [comment]
An interim release is a good idea, there are a lot of improvements since 0.9.

**Comment by hcho3** [comment]
Got it, I will do some plumbing in the CI system in the next few days, and then prepare 1.0.0.preview.1 release.

**Comment by hcho3** [comment]
@CodingCat How about 0.100 or 0.95? "Preview" sounds like the 1.0.0 release is just around the corner, but we have quite a few major features (PySpark) on the line.

**Comment by douglasren** [comment]
Does it support weight xgboost ?

**Comment by CodingCat** [comment]
I am not worrying about the impression of 1.0.0 to users

Spark 3.0 preview is releasing in this month, but formal release is next
April (around spark summit) maybe

**Comment by thesuperzapper** [comment]
@CodingCat at least from the point of view of xgboost4j-spark, that 1.0.0 preview won't be useful for most people, as almost no one is running Spark on 2.12. Additionally, you can't easily get a compiled binary as https://spark.apache.org/downloads.html dosen't distribute compiled versions of Spark for 2.12 with the Hadoop binaries included.

**Comment by CodingCat** [comment]
Then we should release nothing?

**Comment by hcho3** [comment]
@CodingCat @thesuperzapper I thought #4574 would allow for compiling XGBoost with both Scala 2.11 and 2.12?

**Comment by trivialfis** [comment]
Removed:

- [ ]  Release Gpu memory after training #4668

I don't think we can get to there right now.

**Comment by jkbradley** [comment]
@thesuperzapper It will be come easier to develop against the Apache Spark master (3.0) branch and Scala 2.12 after Spark releases a 3.0 preview (targeted pretty soon this fall).  I'd expect a much bigger shift to Scala 2.12 in the Spark community after the final 3.0 release (targeted early 2020), but you're right that there isn't a ton of 2.12 usage now.  I created https://github.com/dmlc/xgboost/issues/4926 to solicit discussion around the upcoming Spark release.

**Comment by trams** [comment]
> @CodingCat @thesuperzapper I thought #4574 would allow for compiling XGBoost with both Scala 2.11 and 2.12? In that case, we should compile XGBoost with 2.11 and upload JAR to Maven.

#4574 does not allow to cross compile.
What it allows is for someone to check out the code, manually override scala version and recompile

So someone may compile a jar with 2.11 and upload to Maven
I had a pull request with migration to SBT which would allow to cross compile
I also know the trick how to support a cross compilation in Maven (we used it in our company). I can share if you are interested

**Comment by trivialfis** [comment]
@hcho3 Is it possible to use CPack https://gitlab.kitware.com/cmake/community/wikis/doc/cpack/PackageGenerators#bundle-osx-only for easing the installation for OSX?  Please ignore this comment if it's not possible.

**Comment by douglasren** [comment]
Does it support Multi objective learning?

**Comment by trivialfis** [comment]
@douglasren Sadly no.  Could you start a new issue so we can discuss it?  The term "multi objective" can vary in contexts, like one objective function for multiple outputs or multiple objectives with multiple outputs?

**Comment by EricSpeidel** [comment]
I would like to cast my vote towards an interim release as well.

**Comment by hcho3** [comment]
#5146 fixes #4477.

**Comment by trivialfis** [comment]
Removed:

- [ ] PySpark API support (https://github.com/dmlc/xgboost/issues/3370) (https://github.com/dmlc/xgboost/pull/4656) .

**Comment by TylerADavis** [comment]
An interim release would be great as the macOS installation is still a pain right now

**Comment by dubeyrahul** [comment]
Can we get documented support for learning to rank (pairwise) with XGBoost4J-Spark? Currently, there is no concrete solution to how to specify training data. There's some confusion around partitioning by groupID and training data needing to follow same partition strategy, but it's quite vague.
An example or clear documentation would be really helpful!

**Comment by lucagiovagnoli** [comment]
I'd like to cast my vote to an interim release as well. We're looking forward to the next version mostly for the missing value fix by @cpfarrell (see https://github.com/dmlc/xgboost/pull/4805). 

Is there a time estimate related to the next release (major or interim)? 

PS:  @thesuperzapper  we're using 2.11 and 2.12 and an interim release would be extremely helpful for us

**Comment by trivialfis** [comment]
@hcho3 Can we make create a release branch and have a week or so for testing?

**Comment by hcho3** [comment]
Yes!

**Comment by terrytangyuan** [comment]
@hcho3 In addition to a branch, we can also make an official release candidate on GitHub Releases so that the community can have more confidence to test it as well.

**Comment by lucagiovagnoli** [comment]
This sounds awesome! Really looking forward to the next release. Let me know if we can help. We're definitely going to test it out at Yelp.

**Comment by hcho3** [comment]
I will cut a new branch `release_1.0.0` after https://github.com/dmlc/xgboost/pull/5248 is merged. Thanks everyone for your patience.

**Comment by hcho3** [comment]
Release candidate is now available for Python: https://github.com/dmlc/xgboost/issues/5253

**Comment by hcho3** [comment]
1.0.0 is now out:
```
pip3 install xgboost==1.0.0
```
